oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedMongoDownloadTask.java

```diff
@@ -349,95 +353,132 @@ private void downloadWithNaturalOrdering() throws InterruptedException, TimeoutE
-     * paths (i.e., /a and /a/b).
-     *
-     * @param pathFilters Empty set if path filtering should be disabled, otherwise the paths that should be included
-     *                    in the Mongo query filters
-     */
-    // Package private for testing
-    static Set<String> extractIncludedPaths(List<PathFilter> pathFilters) {
-        // Path filtering is enabled only if there are no excludedPaths.
-        if (pathFilters == null) {
-            return Set.of();
+    static MongoFilterPaths buildMongoFilter(List<PathFilter> pathFilters) {
+        if (pathFilters == null || pathFilters.isEmpty()) {
+            return null;
         }
-        Set<String> includedPaths = new HashSet<>();
-        Set<String> excludedPaths = new HashSet<>();
+        // Merge and deduplicate the included and excluded paths of all filters.
+        // The included paths will also be further de-deuplicated, by removing paths that are children of other path in
+        // the list, that is, by keeping only the top level paths. To make it easy to de-duplicate, we sort the paths
+        // alphabetically.
+        // The list of excluded paths does not need to be sorted, but we do it anyway for consistency when logging the
+        // list of paths.
+        TreeSet<String> sortedCandidateIncludedPaths = new TreeSet<>();
+        TreeSet<String> candidateExcludedPaths = new TreeSet<>();
         for (PathFilter pathFilter : pathFilters) {
-            includedPaths.addAll(pathFilter.getIncludedPaths());
-            excludedPaths.addAll(pathFilter.getExcludedPaths());
+            sortedCandidateIncludedPaths.addAll(pathFilter.getIncludedPaths());
+            candidateExcludedPaths.addAll(pathFilter.getExcludedPaths());
         }
-        // Sort by natural order, so that parent paths appear before any children. This makes it easier to compute the
-        // common ancestors of included paths
-        List<String> sortedIncludedPaths = includedPaths.stream()
-                .sorted()
-                .collect(Collectors.toList());
-
-        LOG.info("Paths considered for regex filtering. IncludedPaths: {}, excludedPaths: {}", sortedIncludedPaths, excludedPaths);
-        if (sortedIncludedPaths.contains("/")) {
-            LOG.info("Disabling regex path filtering because root path is in the includedPaths: {}", sortedIncludedPaths);
-            return Set.of();
+        // Keep only unique included paths. That is, if paths "/a/b" and "/a/b/c" are both in the list, keep only "/a/b"
+        ArrayList<String> finalIncludedPathsRoots = new ArrayList<>();
+        LOG.debug("Candidate included paths: {}, candidate excluded paths: {}", sortedCandidateIncludedPaths, candidateExcludedPaths);
+        // The quadratic algorithm below could easily be replaced by linear time algorithm, but it's not worth the added
+        // complexity because the list of paths should never be more than a few tens
```

Comment:
Features are often used in surprising ways... What if we don't filter if the list is larger than a 100? With 100 times 100, it would be 10'000, which should still be OK. But above that, it might not.

Reply:
I added a configuration set a limit on the number of paths that can be used to build regex path filters. By default, the value is 20. This is limit is applied after merging and de-duplicating the paths in all the path filters, so the algorithm to de-duplicate the paths might still be exposed to the quadratic complexity when using many paths. I tested in my M1 mac, with 10k unique paths it takes two seconds to de-duplicate the list of paths. I think it's ok, it would be very hard to have a index definition with more than 10k paths, probably other parts of the system would break before getting to this point.